### DISCUSSION TITLE
[Implemented]Plot concept

### DISCUSSION THREAD

**Discussion by PavelBal** [Open-Systems-Pharmacology/OSPSuite-R#601] [Ideas]
Now that we have the `DataSet` class, it is time to implement a concept of creating some figures using `DataSet` and simulated results (https://github.com/Open-Systems-Pharmacology/OSPSuite-R/issues/229). The proposed idea is to have a class that will group multiple data sets and serve as a convenient interface to the TLF library.

For now, I will name the proposed class `Plot`. The workflow would be:
 - Create a `Plot` object
 - Add data to plot - either `DataSet` or from simulation results
 - (Optionally) define which data sets belong together (typically mapping between simulation results and observed data, but not necessarily)
 - (Optionally) specify colors, line types, etc.
 - Set captions, x- and y-axis
 - Create a figure
   - Could be a method of the `Plot` object. In this case, `Plot` could have a property `plotType` (e.g. "individualTimeProfile", "populationQuantiles", "BoxPlot", "PredictedVsObserved"), and a method `Plot$plot()` that executes the specified plot
   - Or functions like `plotIndividualProfile(plot, configuration) `, `plotBoxPlot(plot, configuration` etc.
   - (I prefer the first version - `$plot()` method of the class)

Internally, calling the method `$plot()` (or one of the `plotXXX`-functions) would create a data frame for all data sets (including conversion between units etc), populating the meta data, and pass it to TLF.

## Class fields
- dataSets - returns all data sets. Read-only
- xLim, yLim - limits of the axis (support for multiple y-axis can be added later)
- xLab, yLab - labels of the axis
- xUnit, yUnits - units of the plotted values. The values of `DataSet` objects will be transformed to the selected unit.
- groupings

## Methods/functions for:
- Add `DataSet` objects
- Add simulation results (OutputValues, paths) - creates `DataSet` from for the given paths
- Remove data set (removes by a name/id)

For each `DataSet` in the `Plot` object, following properties are assigned (and can be changed by the user): 
- dataType - observed, simulated, undefined. Adding a DataSet automatically sets its type to "observed", adding a simulation results automatically sets its type to "Simulated". The type is used to decide whether to plot the results as symbols or lines
- Group: name of the group a data set belongs to. Can be used e.g. in the legend. Data sets assigned to the same group will have similar properties like color etc. Group name could also be part of the method signature for adding a data set/simulated result. Data sets without a group a possible
- x/yFactor - x or y values of the respective data set will be multiplied by the factor
- x/yOffset - the value will be added to the x or y values of the respective data set

## Required plot types
  - Individual time-values profile
  - Population time-values profile (quantiles can be defined by the user)
  - BoxPlot
  - Predicted-vs-Observed
  - Multi-panel plots. The panels could be of different types (e.g. Panel a is individual time-values profile, Panel b a pred-vs-observed figure).

@msevestre @abdullahhamadeh @pchelle Would be happy to discuss it with you.

**Comment by PavelBal** [comment]
Following discussion on 01.09.

Separate data grouping from potting. The most annoying and time consuming part in setting up the figures is is definition of what data to use and how these are mapped (e.g. observed data with simulation results). Therefore, I propose the class `DataGrouping` instead of `Plot` with following properties:

## Class fields
- dataSets - returns all data sets. Read-only
- groupings: name of the groupings and the names of data sets assigned to the grouping

## Methods/functions for:
- Add `DataSet` objects
- Add simulation results (OutputValues, paths) - creates `DataSet` from for the given paths
- Remove data set (removes by a name/id)

For each `DataSet` in the `DataGrouping` object, following properties are assigned (and can be changed by the user): 
- dataType - observed, simulated, undefined. Adding a DataSet automatically sets its type to "observed", adding a simulation results automatically sets its type to "Simulated". The type is used to decide whether to plot the results as symbols or lines or for calculating the error etc.
- Group: name of the group a data set belongs to. Can be used e.g. in the legend. Data sets assigned to the same group will have similar properties like color etc. Group name could also be part of the method signature for adding a data set/simulated result. Data sets without a group a possible
- x/yFactor - x or y values of the respective data set will be multiplied by the factor
- x/yOffset - the value will be added to the x or y values of the respective data set

Such a `DataGrouping` can then be used to create standard plots with functions like
- plotIndividualTimeProfile (or just plotTimeProfile, individual/population can be defined in a configuration?)
- plotPredictedVsObserverd
- etc

Each function an also optionally receive a configuration object, having different configurations for different types of plots. Each plot-function can also create a default configuration if none is provided.

The workflow would be:
 - Create a `DataGrouping` object
 - Add data - either `DataSet` or from simulation results
 - (Optionally) define which data sets belong together (typically mapping between simulation results and observed data, but not necessarily)
 - (Optionally) define offsets/factors for the data sets
 - Use the `DataGrouping` with one of the functions (plotTimeProfile, plotPredVsObs, plotBoxPlot, calculateError or whatever...)
 - The configurations for the different functions depend on what the user could set - e.g. line types for time profiles, quantiles for populations, fold-distances for predVsObs...

**Comment by PavelBal** [comment]
@msevestre Did you reach out to @abdelr and @abdullahhamadeh to schedule a meeting?

**Reply by msevestre** [reply]
@PavelBal Yes. @pchelle Already worked on a first draft of the implementation using tlf
